src/cargo/ops/cargo_rustc/fingerprint.rs

```diff
@@ -57,6 +57,15 @@ pub fn prepare_target<'a, 'cfg>(cx: &mut Context<'a, 'cfg>,
     let compare = compare_old_fingerprint(&loc, &*fingerprint);
     log_compare(unit, &compare);
 
+    if compare.is_err() {
+        let source_id = unit.pkg.package_id().source_id();
+        let sources = cx.packages.sources();
+        let source = try!(sources.get(source_id).chain_error(|| {
+            internal("missing package source")
+        }));
+        try!(source.verify(unit.pkg.package_id()));
+    }
```

Comment:
Yeah sure. If `compare.is_err()` is true, it means we're about to rebuild the crate. This typically happens due to modifications to the source or just because it wasn't built previously. In this case we verify the crate source itself, and here is where the directory source will verify all sha256 checksums.